File: live402/batch_profiles/algorand_charge.py
"""Native Algorand MPP charge quote; not sessions or settlement authority."""
import base64, hashlib
from live402 import route_binding as rb
from . import algorand_generic as algo
from .algorand import _address
PROFILE = "algorand-mpp-charge-v1"
NETWORK, ASSET = algo.NETWORK, algo.ASSET
KEYS = {"network","asset","recipient","realm","max_amount_atomic","max_network_fee_micro_algo","fee_payer"}
def check(ok):
    if not ok: raise rb.BindingError("unsupported_algorand_charge")
def keys(value, required, optional=()):
    check(type(value) is dict and set(required)<=set(value)<=set(required)|set(optional))
def uint(value):
    from .base import uint as parse
    return parse(value)
def text(value, maximum):
    check(type(value) is str and 0<len(value.encode())<=maximum)
def address(value):
    _address(value)
    check(value!="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ")
def validate_limits(limits):
    keys(limits,KEYS);check(limits["network"]==NETWORK and limits["asset"]==ASSET)
    address(limits["recipient"]);text(limits["realm"],256)
    check(limits["realm"].strip() and all(32<=ord(c)<127 for c in limits["realm"]))
    uint(limits["max_amount_atomic"]);uint(limits["max_network_fee_micro_algo"])
    if limits["fee_payer"] is not None:
        address(limits["fee_payer"]);check(limits["fee_payer"]!=limits["recipient"])

def fee_model(request):
    """Exact allowed SDK0.9.4 fields; Ed25519 wrapper adds75msgpack bytes.
    No zero-address/default fields or optional signing mechanisms are permitted.
    """
    md=request["methodDetails"];p=md["suggestedParams"];sponsored=md.get("feePayer") is True
    def integer(n): return 1 if n<=127 else 2 if n<=255 else 3 if n<=65535 else 5 if n<=2**32-1 else 9
    def string(s):
        n=len(s.encode());return n+(1 if n<32 else 2 if n<=255 else 3)
    def binary(n): return n+(2 if n<=255 else 3)
    note="mppx:"+md["challengeReference"]+(":"+request["externalId"] if request.get("externalId") else "")
    def size(kind,fee):
        fields={"fv":integer(p["firstValid"]),"lv":integer(p["lastValid"]),"gen":string("mainnet-v1.0"),"gh":binary(32),"grp":binary(32),"snd":binary(32),"type":string(kind)}
        if fee: fields["fee"]=integer(fee)
        if kind=="pay": fields["rcv"]=binary(32)
        else: fields.update({"arcv":binary(32),"xaid":integer(int(md["asaId"])),"aamt":integer(int(request["amount"])),"lx":binary(32),"note":binary(len(note.encode()))})
        return (1 if len(fields)<16 else 3)+sum(string(k)+v for k,v in fields.items())
    kinds=["pay","axfer"] if sponsored else ["axfer"]
    required=lambda n:max(p["minFee"],p["fee"]*n)
    quoted=sum(required(size(k,0)) for k in kinds)
    sizes=[size(k,quoted if i==0 else 0) for i,k in enumerate(kinds)]
    check(quoted>=sum(required(n+75) for n in sizes))
    return quoted,sizes

def validate(request, context, limits):
    validate_limits(limits)
    check(context == rb.request_context(context["url"], "GET"))
    keys(request,{"amount","currency","recipient","methodDetails"},{"description","externalId"})
    text(request["currency"],64);check(request["recipient"]==limits["recipient"] and uint(request["amount"])<=uint(limits["max_amount_atomic"]))
    for key,maximum in [("description",4096),("externalId",256)]:
        if key in request: text(request[key],maximum)
    md=request["methodDetails"]
    keys(md,{"network","asaId","challengeReference","lease","suggestedParams"},{"feePayer","feePayerKey"})
    check(md["network"]==NETWORK and md["asaId"]==ASSET)
    text(md["challengeReference"],256)
    lease=base64.b64encode(hashlib.sha256(md["challengeReference"].encode()).digest()).decode()
    check(md["lease"]==lease)
    sponsored=limits["fee_payer"] is not None
    if sponsored: check(md.get("feePayer") is True and md.get("feePayerKey")==limits["fee_payer"])
    else: check(("feePayer" not in md or md["feePayer"] is False) and "feePayerKey" not in md)
    p=md["suggestedParams"]
    keys(p,{"fee","firstValid","lastValid","genesisHash","genesisId","minFee"})
    check(p["genesisHash"]==NETWORK.split(":",1)[1] and p["genesisId"]=="mainnet-v1.0")
    for key in ["fee","firstValid","lastValid","minFee"]: check(type(p[key]) is int and 0<=p[key]<=2**53-1)
    check(p["minFee"]>=1000 and 0<p["firstValid"]<=p["lastValid"]<=p["firstValid"]+1000)
    count=2 if sponsored else 1
    check(p["minFee"]*count<=uint(limits["max_network_fee_micro_algo"]))
    fee,_=fee_model(request);check(fee<=uint(limits["max_network_fee_micro_algo"]))
    return {"protocol":"mpp","method":"algorand","intent":"charge","network":NETWORK,"asset":ASSET,"recipient":request["recipient"],"amount_atomic":request["amount"],"currency_label":request["currency"],"fee_payer":limits["fee_payer"],"transaction_count":count,"minimum_group_fee_micro_algo":str(p["minFee"]*count),"network_fee_micro_algo":str(fee),"fee_quote_requires_buyer_validation":True,"challenge_reference":md["challengeReference"],"lease":lease,"suggested_params":p}

File: sdk/route-guard/batch-profiles/algorand-charge.mjs
import { createHash } from "node:crypto";

const check = (x) => {
  if (!x) throw Error("unsupported_algorand_charge");
};
const object = (x) => x !== null && typeof x === "object" && !Array.isArray(x);
const keys = (v, required, optional = []) =>
  check(
    object(v) &&
      required.every((k) => Object.hasOwn(v, k)) &&
      Object.keys(v).every((k) => [...required, ...optional].includes(k)),
  );
const text = (v, max) =>
  check(
    typeof v === "string" &&
      Buffer.byteLength(v) > 0 &&
      Buffer.byteLength(v) <= max,
  );
const uint = (v) => {
  check(
    typeof v === "string" &&
      /^[1-9][0-9]{0,19}$/.test(v) &&
      BigInt(v) <= 2n ** 64n - 1n,
  );
  return BigInt(v);
};
function address(v) {
  check(
    typeof v === "string" &&
      /^[A-Z2-7]{58}$/.test(v) &&
      v !== "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ",
  );
  let bits = 0,
    value = 0,
    out = [];
  const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
  for (const c of v) {
    value = (value << 5) | alphabet.indexOf(c);
    bits += 5;
    if (bits >= 8) {
      bits -= 8;
      out.push((value >> bits) & 255);
    }
  }
  check(bits === 2 && (value & 3) === 0);
  const b = Buffer.from(out);
  check(
    b.length === 36 &&
      b
        .subarray(32)
        .equals(
          createHash("sha512-256")
            .update(b.subarray(0, 32))
            .digest()
            .subarray(-4),
        ),
  );
}
export const ALGORAND_CHARGE_PROFILE = "algorand-mpp-charge-v1";
export const ALGORAND_CHARGE_NETWORK =
  "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
export const ALGORAND_CHARGE_ASSET = "31566704";
export const ALGORAND_CHARGE_KEYS = [
  "network",
  "asset",
  "recipient",
  "realm",
  "max_amount_atomic",
  "max_network_fee_micro_algo",
  "fee_payer",
];
export function validateAlgorandChargeLimits(l) {
  keys(l, ALGORAND_CHARGE_KEYS);
  check(
    l.network === ALGORAND_CHARGE_NETWORK && l.asset === ALGORAND_CHARGE_ASSET,
  );
  address(l.recipient);
  text(l.realm, 256);
  check(l.realm.trim().length > 0 && /^[\x20-\x7e]+$/.test(l.realm));
  uint(l.max_amount_atomic);
  uint(l.max_network_fee_micro_algo);
  if (l.fee_payer !== null) {
    address(l.fee_payer);
    check(l.fee_payer !== l.recipient);
  }
}

export function algorandChargeFeeModel(r) {
  const m = r.methodDetails,
    p = m.suggestedParams,
    sponsored = m.feePayer === true;
  const integer = (n) => {
    n = BigInt(n);
    return n <= 127n
      ? 1
      : n <= 255n
        ? 2
        : n <= 65535n
          ? 3
          : n <= 4294967295n
            ? 5
            : 9;
  };
  const str = (s) => {
    const n = Buffer.byteLength(s);
    return n + (n < 32 ? 1 : n <= 255 ? 2 : 3);
  };
  const bin = (n) => n + (n <= 255 ? 2 : 3),
    note =
      "mppx:" + m.challengeReference + (r.externalId ? ":" + r.externalId : "");
  const size = (kind, fee) => {
    const f = {
      fv: integer(p.firstValid),
      lv: integer(p.lastValid),
      gen: str("mainnet-v1.0"),
      gh: bin(32),
      grp: bin(32),
      snd: bin(32),
      type: str(kind),
    };
    if (fee > 0n) f.fee = integer(fee);
    if (kind === "pay") f.rcv = bin(32);
    else
      Object.assign(f, {
        arcv: bin(32),
        xaid: integer(m.asaId),
        aamt: integer(r.amount),
        lx: bin(32),
        note: bin(Buffer.byteLength(note)),
      });
    return (
      (Object.keys(f).length < 16 ? 1 : 3) +
      Object.entries(f).reduce((sum, [k, n]) => sum + str(k) + n, 0)
    );
  };
  const required = (n) => {
      const v = BigInt(p.fee) * BigInt(n);
      return v > BigInt(p.minFee) ? v : BigInt(p.minFee);
    },
    kinds = sponsored ? ["pay", "axfer"] : ["axfer"];
  const quoted = kinds.reduce((n, k) => n + required(size(k, 0n)), 0n),
    sizes = kinds.map((k, i) => size(k, i === 0 ? quoted : 0n));
  check(quoted >= sizes.reduce((n, size) => n + required(size + 75), 0n));
  return { fee: String(quoted), unsignedSizes: sizes };
}

export function validateAlgorandChargeProfile(r, ctx, l) {
  validateAlgorandChargeLimits(l);
  keys(ctx, ["url", "method", "body_sha256"]);
  check(
    ctx.method === "GET" &&
      ctx.body_sha256 === createHash("sha256").update("").digest("hex"),
  );
  keys(
    r,
    ["amount", "currency", "recipient", "methodDetails"],
    ["description", "externalId"],
  );
  text(r.currency, 64);
  check(
    r.recipient === l.recipient && uint(r.amount) <= uint(l.max_amount_atomic),
  );
  for (const [k, n] of [
    ["description", 4096],
    ["externalId", 256],
  ])
    if (Object.hasOwn(r, k)) text(r[k], n);
  const m = r.methodDetails;
  keys(
    m,
    ["network", "asaId", "challengeReference", "lease", "suggestedParams"],
    ["feePayer", "feePayerKey"],
  );
  check(
    m.network === ALGORAND_CHARGE_NETWORK && m.asaId === ALGORAND_CHARGE_ASSET,
  );
  text(m.challengeReference, 256);
  const lease = createHash("sha256")
    .update(m.challengeReference)
    .digest("base64");
  check(m.lease === lease);
  const sponsored = l.fee_payer !== null;
  if (sponsored) check(m.feePayer === true && m.feePayerKey === l.fee_payer);
  else
    check(
      (!Object.hasOwn(m, "feePayer") || m.feePayer === false) &&
        !Object.hasOwn(m, "feePayerKey"),
    );
  const p = m.suggestedParams;
  keys(p, [
    "fee",
    "firstValid",
    "lastValid",
    "genesisHash",
    "genesisId",
    "minFee",
  ]);
  check(
    p.genesisHash === ALGORAND_CHARGE_NETWORK.slice(9) &&
      p.genesisId === "mainnet-v1.0",
  );
  for (const k of ["fee", "firstValid", "lastValid", "minFee"])
    check(Number.isSafeInteger(p[k]) && p[k] >= 0);
  check(
    p.minFee >= 1000 &&
      p.firstValid > 0 &&
      p.firstValid <= p.lastValid &&
      p.lastValid <= p.firstValid + 1000,
  );
  const count = sponsored ? 2 : 1,
    total = BigInt(p.minFee) * BigInt(count);
  check(total <= uint(l.max_network_fee_micro_algo));
  const fee = algorandChargeFeeModel(r).fee;
  check(BigInt(fee) <= uint(l.max_network_fee_micro_algo));
  return {
    protocol: "mpp",
    method: "algorand",
    intent: "charge",
    network: ALGORAND_CHARGE_NETWORK,
    asset: ALGORAND_CHARGE_ASSET,
    recipient: r.recipient,
    amount_atomic: r.amount,
    currency_label: r.currency,
    fee_payer: l.fee_payer,
    transaction_count: count,
    minimum_group_fee_micro_algo: String(total),
    network_fee_micro_algo: fee,
    fee_quote_requires_buyer_validation: true,
    challenge_reference: m.challengeReference,
    lease,
    suggested_params: p,
  };
}
